### DISCUSSION TITLE
Support for external VectorDB

### DISCUSSION THREAD

**Discussion by chrbrnracn** [open-webui/open-webui#938] [Ideas]
In advance: I'm in no means expert for open-webui, so take my quotes with a grain of salt.

Currently open-webui's internal RAG system uses an internal ChromaDB (according to Dockerfile and backend/requirements.txt).

Supporting an external vector database would bring several benefits:
- Ability to use an external process (eg. LlamaIndex) to populate the database.
- Scalability options (ie. running the DB on a different host)
- Could be used my multiple open-webui instances
- Flexibility to add more database products later (eg. pgvector)

**Comment by flefevre** [comment]
I totaly agréé with your proposal. It Could be very usefull if user Could set up an ui property to configuré links to multiple chromadb ensuring sécurity accès and multiple rag options.

Is there àny hint on how to move to thèse tumype of solution ?
Where the code should be modify ?

**Comment by svaeyens** [comment]
+1 for pgvector

**Reply by mulkproject** [reply]
+100 for pgvector

**Reply by EFLKumo** [reply]
+100000000 for pgvector @tjbck 😄

**Reply by jdadame** [reply]
Its already a feature, just use the environmental variables
https://docs.openwebui.com/getting-started/env-configuration/#rag

in docker compose there go under 
```
    environment:
      - 'CHROMA_HTTP_HOST=ip/hostname'
      - 'CHROMA_HTTP_PORT=8000'
```
or put them in your .env file


I have not fully tested this and plan to so please read all documentation linked.

**Reply by EFLKumo** [reply]
Though, it uses chromadb as the vector database client, so I can't migrate vector database with my pgdb.

**Reply by bmdell** [reply]
what about connecting to chromdb that is already pre-populated?

**Reply by justinh-rahb** [reply]
@bmdell WebUI won't use a Chroma database that already has content not added by it, it will error.

**Reply by tan-yong-sheng** [reply]
+1000 for pgvector... Looking forward to it

**Reply by hasani114** [reply]
> Its already a feature, just use the environmental variables https://docs.openwebui.com/getting-started/env-configuration/#rag
> 
> in docker compose there go under
> 
> ```
>     environment:
>       - 'CHROMA_HTTP_HOST=ip/hostname'
>       - 'CHROMA_HTTP_PORT=8000'
> ```
> 
> or put them in your .env file
> 
> I have not fully tested this and plan to so please read all documentation linked.

This doesn't work for me, I tried :

`docker run -d -p 3001:8080 -v open-webui2:/app/backend/data -v D:/open-webui/rag-data:/app/backend/data/docs -e CHROMA_HTTP_HOST="http://host.docker.internal" -e CHROMA_HTTP_PORT=8000 --name open-webui2 --restart always ghcr.io/open-webui/open-webui:main`

And I get this error:

```
[`2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/chromadb/api/base_http_client.py", line 97, in _raise_chroma_error
2024-08-18 18:57:26     raise (Exception(resp.text))
2024-08-18 18:57:26 Exception: <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
2024-08-18 18:57:26 <html><head>
2024-08-18 18:57:26 <title>503 Service Unavailable</title>
2024-08-18 18:57:26 </head><body>
2024-08-18 18:57:26 <h1>Service Unavailable</h1>
2024-08-18 18:57:26 <p>The server is temporarily unable to service your
2024-08-18 18:57:26 request due to maintenance downtime or capacity
2024-08-18 18:57:26 problems. Please try again later.</p>
2024-08-18 18:57:26 </body></html>
2024-08-18 18:57:26 
2024-08-18 18:57:26 
2024-08-18 18:57:26 During handling of the above exception, another exception occurred:
2024-08-18 18:57:26 
2024-08-18 18:57:26 Traceback (most recent call last):
2024-08-18 18:57:26   File "/usr/local/bin/uvicorn", line 8, in <module>
2024-08-18 18:57:26     sys.exit(main())
2024-08-18 18:57:26              ^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
2024-08-18 18:57:26     return self.main(*args, **kwargs)
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1078, in main
2024-08-18 18:57:26     rv = self.invoke(ctx)
2024-08-18 18:57:26          ^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
2024-08-18 18:57:26     return ctx.invoke(self.callback, **ctx.params)
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/click/core.py", line 783, in invoke
2024-08-18 18:57:26     return __callback(*args, **kwargs)
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/uvicorn/main.py", line 410, in main
2024-08-18 18:57:26     run(
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/uvicorn/main.py", line 578, in run
2024-08-18 18:57:26     server.run()
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 61, in run
2024-08-18 18:57:26     return asyncio.run(self.serve(sockets=sockets))
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/asyncio/runners.py", line 190, in run
2024-08-18 18:57:26     return runner.run(main)
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/asyncio/runners.py", line 118, in run
2024-08-18 18:57:26     return self._loop.run_until_complete(task)
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/uvicorn/server.py", line 68, in serve
2024-08-18 18:57:26     config.load()
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/uvicorn/config.py", line 473, in load
2024-08-18 18:57:26     self.loaded_app = import_from_string(self.app)
2024-08-18 18:57:26                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
2024-08-18 18:57:26     module = importlib.import_module(module_str)
2024-08-18 18:57:26              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
2024-08-18 18:57:26     return _bootstrap._gcd_import(name[level:], package, level)
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
2024-08-18 18:57:26   File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
2024-08-18 18:57:26   File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
2024-08-18 18:57:26   File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
2024-08-18 18:57:26   File "<frozen importlib._bootstrap_external>", line 940, in exec_module
2024-08-18 18:57:26   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
2024-08-18 18:57:26   File "/app/backend/main.py", line 30, in <module>
2024-08-18 18:57:26     from apps.socket.main import app as socket_app, get_event_emitter, get_event_call
2024-08-18 18:57:26   File "/app/backend/apps/socket/main.py", line 5, in <module>
2024-08-18 18:57:26     from apps.webui.models.users import Users
2024-08-18 18:57:26   File "/app/backend/apps/webui/models/users.py", line 9, in <module>
2024-08-18 18:57:26     from apps.webui.internal.db import Base, JSONField, Session, get_db
2024-08-18 18:57:26   File "/app/backend/apps/webui/internal/db.py", line 7, in <module>
2024-08-18 18:57:26     from apps.webui.internal.wrappers import register_connection
2024-08-18 18:57:26   File "/app/backend/apps/webui/internal/wrappers.py", line 9, in <module>
2024-08-18 18:57:26     from config import SRC_LOG_LEVELS
2024-08-18 18:57:26   File "/app/backend/config.py", line 1074, in <module>
2024-08-18 18:57:26     CHROMA_CLIENT = chromadb.HttpClient(
2024-08-18 18:57:26                     ^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/chromadb/__init__.py", line 204, in HttpClient
2024-08-18 18:57:26     return ClientCreator(tenant=tenant, database=database, settings=settings)
2024-08-18 18:57:26            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/chromadb/api/client.py", line 61, in __init__
2024-08-18 18:57:26     self._validate_tenant_database(tenant=tenant, database=database)
2024-08-18 18:57:26   File "/usr/local/lib/python3.11/site-packages/chromadb/api/client.py", line 380, in _validate_tenant_database
2024-08-18 18:57:27 Loading WEBUI_SECRET_KEY from file, not provided as an environment variable.
2024-08-18 18:57:27 Loading WEBUI_SECRET_KEY from .webui_secret_key
2024-08-18 18:57:26     raise ValueError(
2024-08-18 18:57:26 ValueError: Could not connect to tenant default_tenant. Are you sure it exists?`](url)
```

**Reply by diegossalmeida** [reply]
@hasani114 If you're still experiencing issues, try setting CHROMA_HTTP_HOST=host.docker.internal (without the http:// prefix).

Worked for me.

**Comment by Manfredzimmermann** [comment]
There must be a way to connect Open Web UI to an external Vector database! What would be very cool is if you could select an external Vector database under Settings in Open Web UI. How can such a functionality be built into the settings? Simply add a button, such as "select a Vector database" or "add Vector database". Then you can enter the location of the vector database and other parameters and you're done.

**Comment by Manfredzimmermann** [comment]
How can I connect the Vespa vector database with Open Webui?

**Reply by EFLKumo** [reply]
no way. Now it only supports chromadb. However you can write a bridge between chromadb and vespa, though it doesn't look classy

**Comment by flefevre** [comment]
Did you set up your own chroma instance?
If yes, did you expose the port?

More over the chroma client and server should have the same version

See here

https://github.com/chroma-core/chroma/issues/2510



Le dim. 18 août 2024, 16:00, Hasan Iqbal ***@***.***> a
écrit :

**Reply by hasani114** [reply]
I checked, they are the same version 0.5.8 on the openwu server and the docker container hosting chromadb (I checked using the /docs url to check the version.

I also downloaded chromadb admin to see if there was something wrong with the container access and nope, it worked perfectly using the http://host.docker.internal:8000 url.

previously, I was using vector admin (another GUI to manage vector databases) and it worked perfectly with my chroma instance. 

I also created a new chroma container to ensure it was a fresh install and had nothing from before (based on the comment above about open webui not supporting dbs that already have collections in them.

**Comment by Manfredzimmermann** [comment]
I have experimented a lot with RAG and ChromaDB under Open WEBUI. The results were not good! The whole thing is not ready for production yet. I read somewhere that RAG2 is coming. I hope RAG2 is better!

**Comment by flefevre** [comment]
Could you please tell us which vector database to use?
What are the better parameters for chunck etc... For basic scientific pdf
publication?

Thanks

Le lun. 19 août 2024, 12:49, Manfredzimmermann ***@***.***> a
écrit :

**Reply by Manfredzimmermann** [reply]
I have used different chunk methods under ChromaDB. There is a fixed size and recursive method and so on. But the results were not good! The whole thing is not suitable for a productive environment, in my opinion.

**Comment by flefevre** [comment]
Thanks for sharing your experience.
Do you have any advice?
To switch to another vector database?
Any reference on rag2?

Thanks again

Le lun. 19 août 2024, 21:11, Manfredzimmermann ***@***.***> a
écrit :

**Comment by Manfredzimmermann** [comment]
I can't give any advice at the moment without a clear conscience! Unfortunately, I haven't found any information on the Internet yet about how to use RAG2 with Python and the like. Open-webui is cool and remarkable, but not yet suitable for a productive environment. But I'm sure that it will be in the future if programming continues. It's missing some features or functions. Such as connection to an LDAP server, assignment of more than one LLM model to a specific group. So, different user groups with different LLMs. RAG2 or something similar that is able to write various documents (PDF, CSV, Word, Excel and etc.) very well into a vector database and delivers results for queries with 99% accuracy. When it comes to medical data, then 100% accuracy. What would also be very cool is chatting with structured data, i.e. a connection to an SQL database, so that a user can chat with the SQL database using natural language and also create diagrams if necessary. But the LLM must be able to interpret the queries very well!

**Comment by tazomatalax** [comment]
You can use a N8N webhook function to chat with an AI agent connected to an external vector database. Just not sure the best way to seamlessly upload documents to this database, via Open WebUI --> N8N --> write to vector database.

Any Ideas?

**Comment by tazomatalax** [comment]
It seems you can easily add your own qdrant vector store!

just set 
```
VECTOR_DB=qdrant 
QDRANT_URI=http://qdrant:6333/
```

in the docker-compose.yml

you will also need to have qdrant defined as a service in the docker-compose.yml something like this (with network and volumes defined)

```
  qdrant:
    image: qdrant/qdrant
    container_name: qdrant
    networks: ['ai-network']
    restart: unless-stopped
    ports:
      - 6333:6333
    volumes:
      - qdrant_storage:/qdrant/storage

**Reply by jaespinozasj** [reply]
The problem is that when uploading documents to knowledge in openwebui, they are not added to qdrant, at least that happened to me, what is your experience, could you provide a guide if it worked for you?

**Reply by tazomatalax** [reply]
Did you set 
```
environment:
    - VECTOR_DB=qdrant 
    - QDRANT_URI=http://qdrant:6333/
```

in your Open WebUI service of docker compose?

**Reply by bobybonillamx** [reply]
Hi! i've tryied this out but i can't figure out what is wrong 

I first try what you said and didn't work, then i saw in the Open WebUI documentation (https://docs.openwebui.com/getting-started/env-configuration/#qdrant)

That it need an API key i have the service in the config file like this:

` qdrant:
    image: qdrant/qdrant
    container_name: qdrant
    networks: ['demo']
    restart: unless-stopped
    ports:
      - 6333:6333
    volumes:
      - qdrant_storage:/qdrant/storage`

and the enviroment like this:

`"QDRANT_URI=http://host.docker.internal:6333",
"VECTOR_DB=qdrant",
"QDRANT_API_KEY=API_KEY"
`

i've tryied with http://localhost:6333, http://qdrant:6333/, and the host docker internal woth and without the last / but nothing seems to work

in the console or logs i got nothing about Qdrant conection, how can i verify?

Thanks

**Reply by sudhirvadlamani** [reply]
I don't remember setting any API key while testing with qdrant. Assuming Ollama/OpenwebUI/qdrant is on thee same host. Please run below commands:

-> netstat -anp|grep "6333" ( check output on which IP inteface qdrant is listening)

-> You can run OpenWebUI with option as below:

docker run -d -p 48080:8080 --add-host=host.docker.internal:$IPAddress -e OLLAMA_BASE_URL=http://$IPAddress:11434 -e VECTOR_DB=qdrant -e QDRANT_URI=http://$IPAddress:6333/ -v open-webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:main

**Reply by Yashh56** [reply]
> docker run -d -p 48080:8080 --add-host=host.docker.internal:$IPAddress -e OLLAMA_BASE_URL=http://$IPAddress:11434 -e VECTOR_DB=qdrant -e QDRANT_URI=http://$IPAddress:6333/ -v open-webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:main

I performed this docker command and open webUI is running but how can I verify that it is connected with qdrant db?

**Comment by sudhirvadlamani** [comment]
Just want to know whether RAG with OpenWebUI is working with ChromaDB as an external database. I am facing an error as below which I believe parameters in below json "where":{},"where_document":{} should have values or NULL instead of being empty. Any settings to be changed in "Knowledge Base" of OpenWebUI?

POST /api/v1/collections/$CollectionID/query 
{"query_embeddings":[[xxxxxxxxxxxxxxxxxxxx],"n_results":3,"where":{},"where_document":{},"include":["metadatas","documents","distances"]}

HTTP/1.1 400 Bad Request
{"error":"InvalidArgumentError","message":"Expected where to have exactly one operator, got {}"}

**Comment by flefevre** [comment]
Openwebui is fully interoperable with chromadb. I am using it. Be careful
your chromadb instance should have the same versio ideally with the
chromadb client version in openwebui.

Le jeu. 16 janv. 2025, 18:53, sudhirvadlamani ***@***.***> a
écrit :

**Comment by tazomatalax** [comment]
Load a document and then go to qdrant uri and see if it has been added.

Tasman
________________________________

**Reply by Yashh56** [reply]
Thanks, Man it's working. But I have a question suppose I insert the data in my qdrant database from some other source(not from WebUI) can I access them via WebUI like RAG or something like that?

**Comment by apaul030515** [comment]
So far, does the latest version of OpenWebUI support pgvector vector databases?

**Reply by sudhirvadlamani** [reply]
@apaul030515 : As per documentation it's supported. https://docs.openwebui.com/getting-started/env-configuration
<img width="518" alt="image" src="https://github.com/user-attachments/assets/98093771-4afe-4fd7-9e66-a3862bd9ca0a" />

**Comment by danielrosehill** [comment]
Just a thought (I've noticed that the RAG performance is not quite there but not a million miles off either and FWIW, I would personally love to use Weaviate, whether self-hosted or remote):

 Chroma DB may be perfectly sufficient for the vast majority of users.

But what wwould definitely be useful is some more granular embedding options:

For example, if the same user wishes to embed a big chunk of documents and CSV data ... the same embedding profile isn't going to work well for both of them. 

As far as I can tell, right now, we can only configure one set of RAG variables and they are being applied across the board.

**Reply by Jeevanhm** [reply]
yes, pdf or word documents and CSV data doesn't work well under same embedding profile.  Have you tried changing the profile under advance params on the model used for RAG?

**Reply by sudhirvadlamani** [reply]
@Jeevanhm :  Are you talking about configuration option in "Admin Settings" -> "Documents" section?

**Reply by sudhirvadlamani** [reply]
> Just a thought (I've noticed that the RAG performance is not quite there but not a million miles off either and FWIW, I would personally love to use Weaviate, whether self-hosted or remote):
> 
> Chroma DB may be perfectly sufficient for the vast majority of users.
> 
> But what wwould definitely be useful is some more granular embedding options:
> 
> For example, if the same user wishes to embed a big chunk of documents and CSV data ... the same embedding profile isn't going to work well for both of them.
> 
> As far as I can tell, right now, we can only configure one set of RAG variables and they are being applied across the board.

I was never able to successfully use Weaviate. Tried to connect through Verba with Docker/Custom option and request fails always.

**Comment by ips972** [comment]
switch to postgress databes and pgvector as verctor database. everything works great for 200 users.. 
2 instances on open webui with redis for connection management..too.

**Reply by Jeevanhm** [reply]
hello, do you have the env variable to connect pgvector DB? 

Below is for qdrant:
docker run -d -p 48080:8080 --add-host=host.docker.internal:$IPAddress -e OLLAMA_BASE_URL=http://$IPAddress:11434 -e VECTOR_DB=qdrant -e QDRANT_URI=http://$IPAddress:6333/ -v open-webui:/app/backend/data --name open-webui --restart always ghcr.io/open-webui/open-webui:main

**Reply by SquadUpSquid** [reply]
Hello @ips972, I would also like to know how you got it connected with pgvector. I can't find any other documentation on how to do this. 

I've got the DB working and wrote a pipeline to get the embeddings going just don't know how to connect it all with open-webui..

**Reply by rgaricano** [reply]
env vars:
VECTOR_DB="pgvector"
PGVECTOR_DB_URL="postgresql://openwebui:OpenWebUI@localhost:5432/openwebui_pg"
DATABASE_URL="postgresql://openwebui:OpenWebUI@localhost:5432/openwebui_db"
--------------------user postgresql______J ㅤㅤㅤㅤㅤㅤ|ㅤㅤㅤㅤㅤ|ㅤㅤㅤㅤㅤㅤㅤㅤL_________database
-------------------------------------------------password_____Jㅤㅤㅤㅤㅤ|____postgresql-server--------

**Comment by ips972** [comment]
@SquadUpSquid 

Hi, no problem, glad to help.

create a docker-compose.yml for your services.
/opt/open-webui-stack/docker-compose.yml

add the following to the file:


services:
  #postgres with pgvector included!! 
  postgres:
    image: ankane/pgvector
    container_name: postgres
    environment:
      POSTGRES_USER: user
      POSTGRES_PASSWORD: password
      POSTGRES_DB: openwebui
    volumes:
      - /opt/open-webui-stack/postgres_data/:/var/lib/postgresql/data
    ports:
      - '5432:5432'
    restart: unless-stopped

  openWebUI:
    #    image: ghcr.io/open-webui/open-webui:main  #for CPU baesed
    image: ghcr.io/open-webui/open-webui:cuda       #for GPU accelerated
    container_name: openwebui1
    depends_on:
        - postgres
    volumes:
      - ./data/open-webui:/app/backend/data

    environment:
      #this makes sure that users don't loose login after stack restart
      WEBUI_SECRET_KEY: bf434gf3840f374f3784g34fb34jf3478f3b4f34f34f
      RAG_WEB_SEARCH_TRUST_ENV: True
      GLOBAL_LOG_LEVEL: DEBUG
      DEFAULT_USER_ROLE: pending
      DATABASE_URL: postgresql://user:password@postgres:5432/openwebui
      DATA_DIR: /app/backend/data
      OLLAMA_BASE_URLS: http://ollama:11434
      PGVECTOR_DB_URL: postgresql://user:password@postgres:5432/openwebui
      VECTOR_DB: pgvector
      SOCKET_LOG_LEVEL: DEBUG
      WEBSOCKET_MANAGER: redis
      WEBSOCKET_REDIS_URL: redis://redis1:7000/1

      CORS_ALLOW_ORIGIN: "*" # This is the current Default, will need to change before going live
      RAG_EMBEDDING_MODEL_TRUST_REMOTE_CODE: "True"
      NVIDIA_DRIVER_CAPABILITIES: compute,video,utility
      NVIDIA_VISIBLE_DEVICES: ALL
      AIOHTTP_CLIENT_TIMEOUT: 1200
    ports:
      - 3000:8080
    extra_hosts:
      - host.docker.internal:host-gateway
    restart: unless-stopped
    deploy:
      resources:
        reservations:
          devices:
            - driver: ${OLLAMA_GPU_DRIVER-nvidia}
              count: ${OLLAMA_GPU_COUNT-1}
              capabilities:
                - gpu

  redis1:
    image: redis:latest
    ports:
      - 7000:7000
    networks:
      - redis-cluster-compose
    hostname: redis1
    volumes:
      - ./7000:/redis
      - ./7000:/data
    command: redis-server /redis/redis.conf

  redis-insight:
    image: redislabs/redisinsight
    ports:
      - 5540:5540
      - 8001:8001
    volumes:
      - ./redisinsight:/db



save the file and then run:
docker compose -f docker-compose.yml up -d --build --force-recreate --remove-orphans


best hide open webui port 3000 behind haproxy or nginx for SSL/HTTPS with your internal URL.

**Reply by SquadUpSquid** [reply]
Thanks, it didn't work out of the gate but this helps me move in the right direction!

**Reply by ips972** [reply]
it should work?! i copied most of the docker-compos.yml i'm running in production  .... strange. 

I'm running this setup with 2 instances of open webui behind a LB both working of the redis, postgres and pgvector - all is working GREATE.. 

my next upgrade is to move the local files in open-webui to S3  service - i just need to figure out how to migrate all the files (thousands) and entries currently used - without having to let the users know - they lost it all. :-) LOL

**Reply by SquadUpSquid** [reply]
Yea I'm not super sure why either, it's got to be something simple that I'm missing..

haha, enjoy the migration. I'll post back in here if I have any other questions, thanks again!

**Reply by SquadUpSquid** [reply]
`  File "/app/backend/open_webui/internal/db.py", line 68, in handle_peewee_migration
    if db and not db.is_closed():
       ^^
UnboundLocalError: cannot access local variable 'db' where it is not associated with a value
`

This is the error i'm getting in the logs btw

I meant to add this in the last message.

**Edit: FIXED**
The issue for this was I was setting DATABASE_URL and not PGVECTOR_DB_URL env var. Even though DATABASE_URL is the default for PGVECTOR_DB_URL for some reason it wasn't working.

**Reply by rgaricano** [reply]
no, one is the database for openwebui data (DATABASE_URL) and other is for vectorized info (PGVECTOR_DB_URL), same data server (postgresql), differents databases.

**Reply by tan-yong-sheng** [reply]
> my next upgrade is to move the local files in open-webui to S3 service - i just need to figure out how to migrate all the files (thousands) and entries currently used - without having to let the users know - they lost it all. :-) LOL

Hi @ips972,

Could this be solved via rclone? I personally used rclone to sync / move the file between one cloud (include server's filesystem) and another cloud storage providers... But I never handle such large scale of syncing before.

Website: https://rclone.org/

---

Btw, wanna ask about few things about your setup if you're ok to share:
- how many active users are you supporting (e.g. max concurrent users)
- Are you using docker swarm or Kubernetes to support it if your user base is large. How many openwebui instances or postgres instances you have created to support the concurrent usage of users. Thanks.

**Reply by ips972** [reply]
1. registered about 230 users, as this is in POC  fase. 
2. 30 to 40 connected users, but about one user every 20 minutes use the system. my setup can handle about 20 responses with ease at the same time. tested with 20 chrome windows open , where i asked them to generate a 2000 words bed time stories. had a feeling i can support more. but at the moment . 20 is enough. 
just to clearify, had to switch from ollama to vllm as backend.. as it was able to give .uch higher concurrent performance then ollama on same setup.  

3. using docker,  and docker compose. nothing else.. via cli..  two servers, each doing their thing.  have more features split between them to  spread the load on vram. like seaxng, tika, pipelines, comfyui cintainers. each managed with their own compose.yml service. 

4. running 2 openwebui in the same compose.yml sharing same volume. and working of the same databases and redis.  both behind haproxy loadbalancer with http and web socket supports. 


comphyui running flux1 dev... everyone very happy with it.. 


hope this helps!

**Reply by tan-yong-sheng** [reply]
I see, thanks a lot for the info

**Reply by bylin-tf** [reply]
@ips972 @rgaricano are we required to have two different databases on the same server for pgvector and postgres to work? I have used your docker compose @ips972  and it seems to be working. I can see document chunk in the openwebui db. but I'm confused by @rgaricano referencing 2 different URIs for DATABASE_URL and PGVECTOR_DB_URL .

Have you guys also tried with newer version of postgres and pgvector. ankane is postgres 15 and pgvector 0.5.1. 
pgvector image is on 0.8.1 and postgres 18.

When attaching the knowledge base to the model now it doesnt seem to search the knowledge base correctly. as it return that it doesnt know. 

Any help would be appreciated

**Reply by ips972** [reply]
I'm used to use the ankane is postgres 15 as both database and pgvector. 
have since switched to qdrant cluster for much higher performance with 6 dedicated external VM servers. 
since in the last few releases there have been much improvements to the qdrant support. 
and also created a dedicated VM with postgres - also for beter performance and separation.

**Reply by bylin-tf** [reply]
ah okay we are running a POC on a single node right now so I think sticking ankane or pgvector own images should work fine. But Thanks! @ips972

**Reply by bylin-tf** [reply]
@ips972 can I ask what kind of hardware specs you are running to support that amount of users? Curious about your context length in combination with the amount of files you have in your knowledge base. Again, I appreciate any help

**Reply by ips972** [reply]
@bylin-tf  

4 host servers, each with 96 cores, 1TB RAM , NVME SAN Storage. 
about 20 vms to run the hole operation. (docker swarm and dedicated servers for DB and other services.)
each vm with 32GB ram to 128GB depending on the usage.. 

really large setup.. .

**Comment by tan-yong-sheng** [comment]
To me, it seems like when I set this up, I got an error where /redis/redis.conf is not found. (Note: not fully sure because I not follow exactly as I change a few settings based on my own preference)

```
redis1:
    image: redis:latest
    ports:
        - 7000:7000
    networks:
        - redis-cluster-compose
    hostname: redis1
    volumes:
        - ./7000:/redis
        - ./7000:/data
    command: redis-server /redis/redis.conf
```

**Comment by bytewavenetwork** [comment]
+20 for Milvus and ElasticSearch

**Reply by athoik** [reply]
Please note that, just recently impement support for pinecone. Will be available at next relase 0.6.6

**Comment by lawalselim** [comment]
One challenge I face now is getting to make this work with TimescaleDB. Anyone tried this before? How did you go about it. 
Another one will be support for multiple tables when connecting to a db like timescale.

**Comment by arunbugkiller** [comment]
Hi, I have been using OpenwebUI with chromaDB from past 8 months supporting 500 users with 7 different RAG agents. 

I am planning to shift from ChromaDB to Qdrant. Shall I go ahead and do it? If yes, do I need to re-upload all the documents to knowledge bases.

**Reply by sir3mat** [reply]
How do you handle the rag agents?

**Reply by Jeevanhm** [reply]
@arunbugkiller great I had to reupload all the documents when I moved to Qdrant

**Reply by arunbugkiller** [reply]
@sir3mat as mentioned earlier, I am using Tika, Ollama (nomic-embed-text) and chroma DB to process and store the documents. Currently my source folders are in sharepoint, one-drive. I am using Open-webUI APIs to develop a cron that gets the respective docs from source folders and then upload and subsequently insert these docs in the knowledge base. 

Hope this answers your query.

**Reply by rgaricano** [reply]
if you migrate vectordb it's not necessary to re-upload docs.
I did migration to postgresql (with https://github.com/taylorwilsdon/open-webui-postgres-migration) but I didn't to qdrant, maybe this help to do migration to qdrant: https://github.com/qdrant/migration

**Reply by kkTea** [reply]
sadly, the repo has no examples for pgvector

**Reply by Ahmedayaz1210** [reply]
Hey how did you connect an external chroma db server to open web ui replacing the default one? Right now what I have done is created a network on docker and put both my chroma db server and open web ui in there and connected them but now how do I index data into chroma db that it shows up in the knolwedge base of open web ui or somehow connect it so I can create a rag chatbot?

**Comment by ArcCal** [comment]
fully support the improvement to connect to external vector databases like Qdrant and PgVector via environment variable configuration. This will greatly benefit scalability and production readiness.

**Reply by rgaricano** [reply]
https://docs.openwebui.com/getting-started/env-configuration#vector-database

**Comment by Ahmedayaz1210** [comment]
Hello, I am currently trying to connect external chroma db to open web ui. I was reading threads and saw people have done it before but I am still confused how to do it. Right now what I have done is created a network on docker and put both my chroma db server and open web ui in there and connected them but now how do I index data into chroma db that it shows up in the knolwedge base of open web ui or somehow connect it so I can create a rag chatbot?
